lib/plugin.py

```diff
@@ -70,18 +95,28 @@ def create_push_details(self):
         return {k: v for k, v in push_details if v is not None}
 
     def _provision_user(self):
-        phone_attribute = self.plugin_configuration.get('starling_auto_provision', 'phone_attribute')
-        email_attribute = self.plugin_configuration.get('starling_auto_provision', 'email_attribute')
-        name_attribute = 'displayName'
-        if phone_attribute and email_attribute:
+        if self._ldap_user_info.phone and self._ldap_user_info.email:
             self.logger.debug('Start auto provisioning of user: {}'.format(self.username))
-            client = self.construct_mfa_client()
-            ldap_info = self._query_user_ldap_information((phone_attribute, email_attribute, name_attribute))
-            user_id = client.provision_user(ldap_info[phone_attribute],
-                                            ldap_info[email_attribute],
-                                            ldap_info[name_attribute])
+            cached_user_info = self.__cache.get(self._user_id_cache_key)
+            if cached_user_info:
+                if cached_user_info['is_valid']:
+                    self.mfa_identity = cached_user_info['user_id']
+                    return
+                else:
+                    reason = 'Cached user ID is invalid, try again later; user_id={}'.format(cached_user_info['user_id'])
+                    self.logger.warning(reason)
+                    return AAResponse.deny(reason=reason)
+            user_id = self._client.provision_user(self._ldap_user_info.phone,
+                                                  self._ldap_user_info.email,
+                                                  self._ldap_user_info.name)
+            self.__cache.set(key=self._user_id_cache_key, value={'user_id': user_id, 'is_valid': True}, ttl=0)
             self.mfa_identity = user_id or self.mfa_identity
 
-    def _query_user_ldap_information(self, required_attributes):
+    def _query_user_ldap_information(self):
+        phone_attribute = self.plugin_configuration.get('starling_auto_provision', 'phone_attribute')
+        email_attribute = self.plugin_configuration.get('starling_auto_provision', 'email_attribute')
+        name_attribute = 'displayName'
+        required_attributes = [phone_attribute, email_attribute, name_attribute]
```

Comment:
These are not required, if phone or email attribute is not defined, then we can just fill LDAPInfo with Nones. See default config: "# To turn on auto-provisioning, you have to specify both attributes"

Reply:
Done